### PR TITLE
Use new CV capabilities via arraycontext

### PR DESCRIPTION
**Pull request by MTCam**
It is regretful this is such a giant haul, but it turns out this was something real hard to do incrementally.

### CODE REVIEW

File: mirgecom/fluid.py [hunk 1]
```diff
@@ -313,6 +312,30 @@ def join_conserved(dim, mass, energy, momentum,
     return result
 
 
+def create_conserved(dim, mass, energy, momentum, species_mass=None):
```
Comment (inducer) [inline mirgecom/fluid.py:315]: Why is this needed? Couldn't we just call the `ConservedVars` constructor?

File: mirgecom/euler.py [hunk 2]
```diff
@@ -144,35 +143,28 @@ def euler_operator(discr, eos, boundaries, q, t=0.0):
...
+                                                        eos=eos,
+                                                        btag=btag,
+                                                        t=t,
+                                                        cv=cv),
+                eos=eos
+            )
+            for btag in boundaries)
+    ).join()
+
+    return split_conserved(
```
Comment (inducer) [inline mirgecom/euler.py:166]: Is this `join`/`split` pair here because `inverse_mass` and `face_mass` don't understand array containers? Let's teach them! I feel like we should ultimately be able to fire all uses of `{split,join}_conserved` from the code base.

**Review by inducer — COMMENTED** [review]
Thanks so much for taking this on, this is great! Two thoughts from a quick scroll below.

File: mirgecom/euler.py (hunk 2 above)
Comment (inducer) [reply mirgecom/euler.py:166]: (cc @thomasgibson )

**Comment by inducer** [comment]
And: I'm super supportive of trying to get this in super ASAP.

File: mirgecom/euler.py (hunk 2 above)
Comment (thomasgibson) [reply mirgecom/euler.py:166]: Sounds like a good plan. Let's get https://github.com/inducer/grudge/pull/74 in ASAP so we can build on top of it. I'm keeping it regularly updated with all the new changing parts.

Comment (MTCam) [reply mirgecom/euler.py:166]: Yes, and this one just above:

```python
    vol_weak = discr.weak_div(inviscid_flux(discr=discr, eos=eos, cv=cv).join())
```

File: mirgecom/flux.py [hunk 3]
```diff
@@ -72,6 +72,7 @@ def lfr_flux(q_tpair, flux_func, normal, lam):
         object array of :class:`meshmode.dof_array.DOFArray` with the
         Lax-Friedrichs/Rusanov flux.
     """
-    flux_avg = 0.5*(flux_func(q_tpair.int)
-                    + flux_func(q_tpair.ext))
-    return flux_avg @ normal - 0.5*lam*(q_tpair.ext - q_tpair.int)
+    flux_avg = 0.5*(flux_func(cv_tpair.int)
+                    + flux_func(cv_tpair.ext))
+    return flux_avg.join() @ normal - 0.5*lam*(cv_tpair.ext.join()
+                                               - cv_tpair.int.join())
```
Comment (MTCam) [inline mirgecom/flux.py:78]: These joins are after:

```python
return flux_avg @ normal - 0.5*lam*(cv_tpair.ext - cv_tpair.int)
```

gave this error:

```
E ValueError: matmul: Input operand 0 does not have enough dimensions (has 0, gufunc core with signature (n?,k),(k,m?)->(n?,m?) requires 1)
```

After using the "join" of flux_avg:

```python
return flux_avg.join() @ normal - 0.5*lam*(cv_tpair.ext - cv_tpair.int)
```

this error:

```
E TypeError: unsupported operand type(s) for *: 'DOFArray' and 'ConservedVars'
```

Comment (MTCam) [inline mirgecom/flux.py:78]: These joins are after:

```python
return flux_avg @ normal - 0.5*lam*(cv_tpair.ext - cv_tpair.int)
```

gave this error:

```
E ValueError: matmul: Input operand 0 does not have enough dimensions (has 0, gufunc core with signature (n?,k),(k,m?)->(n?,m?) requires 1)
```

After using the "join" of flux_avg:

```python
return flux_avg.join() @ normal - 0.5*lam*(cv_tpair.ext - cv_tpair.int)
```

this error:

```
E TypeError: unsupported operand type(s) for *: 'DOFArray' and 'ConservedVars'
```

Comment (MTCam) [inline mirgecom/flux.py:78]:
```python
return flux_avg @ normal - 0.5*lam*(cv_tpair.ext - cv_tpair.int)
```

gave this error:

```
E ValueError: matmul: Input operand 0 does not have enough dimensions (has 0, gufunc core with signature (n?,k),(k,m?)->(n?,m?) requires 1)
```

After using the "join" of flux_avg:

```python
return flux_avg.join() @ normal - 0.5*lam*(cv_tpair.ext - cv_tpair.int)
```

this error:

```
E TypeError: unsupported operand type(s) for *: 'DOFArray' and 'ConservedVars'
```

File: mirgecom/fluid.py (hunk 1 above)
Comment (MTCam) [reply mirgecom/fluid.py:315]: Can we overload the constructor and retain the shape and structure validation?

Comment (inducer) [reply mirgecom/fluid.py:315]: Fair point!

There's `__post_init__` in a dataclass ([docs](https://docs.python.org/3/library/dataclasses.html)), but revalidating on every object construction is an easy way to get really slow code. I'm warming to having an explicit function that does this. I've tended to call these `make_xxx`. (examples: `make_kernel` in loopy, `make_xxx` in [pytato](https://documen.tician.de/pytato/reference.html#user-facing-node-creation))

File: mirgecom/flux.py (hunk 3 above)
Comment (inducer) [reply mirgecom/flux.py:78]: Great question! https://github.com/inducer/arraycontext/issues/12

File: mirgecom/fluid.py (hunk 1 above)
Comment (MTCam) [reply mirgecom/fluid.py]: I hope this helps a little. (d012fba)

File: mirgecom/flux.py (hunk 3 above)
Comment (inducer) [reply mirgecom/flux.py:78]: As of https://github.com/inducer/arraycontext/pull/15, this should now be doable. Note that you may need to tweak some of the arguments to `with_container_arithmetic`. See the [tests](https://github.com/inducer/arraycontext/blob/e5ff3c6eaeba199766598384b6b8173f37b38714/test/test_arraycontext.py#L464-L468).

Comment (MTCam) [reply mirgecom/flux.py]: (bc09408)

**Comment by MTCam** [comment]
Ok, I've spent some time rolling these changes all the way up the Y1 tower - and everything only improves with it.  I still regret this PR is so giant, but the changes are mostly trivial - and where they are not trivial they improve the code. 

Hopefully not a terribly painful review.

File: mirgecom/simutil.py [hunk 4]
```diff
@@ -112,8 +110,8 @@ def sim_checkpoint(discr, visualizer, eos, q, vizname, exact_soln=None,
         actx = cv.mass.array_context
         nodes = thaw(actx, discr.nodes())
         expected_state = exact_soln(x_vec=nodes, t=t, eos=eos)
-        exp_resid = q - expected_state
-        err_norms = [discr.norm(v, np.inf) for v in exp_resid]
+        exp_resid = cv - expected_state
+        err_norms = [discr.norm(v, np.inf) for v in exp_resid.join()]
         maxerr = max(err_norms)
```
Comment (inducer) [inline mirgecom/simutil.py:115]:
```suggestion
        maxerr = discr.norm(exp_resid, np.inf)
```

File: test/test_bc.py [hunk 5]
```diff
@@ -88,24 +87,25 @@ def test_slipwall_identity(actx_factory, dim):
...
             bnd_norm = partial(discr.norm, p=np.inf, dd=BTAG_ALL)
 
-            bnd_pair = wall.boundary_pair(discr, uniform_state, t=0.0,
-                                          btag=BTAG_ALL, eos=eos)
-            bnd_cv_int = split_conserved(dim, bnd_pair.int)
-            bnd_cv_ext = split_conserved(dim, bnd_pair.ext)
+            bnd_pair = wall.boundary_pair(discr, btag=BTAG_ALL,
+                                          eos=eos, cv=uniform_state)
+            # bnd_cv_int = split_conserved(dim, bnd_pair.int)
+            # bnd_cv_ext = split_conserved(dim, bnd_pair.ext)
```
Comment (inducer) [inline test/test_bc.py:93]: ?

File: test/test_filter.py [hunk 6]
```diff
@@ -186,7 +186,7 @@ def test_filter_function(actx_factory, dim, order, do_viz=False):
     # the filter unharmed.
     from mirgecom.initializers import Uniform
     initr = Uniform(dim=dim)
-    uniform_soln = initr(t=0, x_vec=nodes)
+    uniform_soln = initr(t=0, x_vec=nodes).join()
```
Comment (inducer) [inline test/test_filter.py:189]: #377 (OK to resolve when seen)

File: test/test_euler.py [hunk 7]
```diff
@@ -225,75 +221,42 @@ def test_inviscid_mom_flux_components(actx_factory, dim, livedim):
...
+        p = eos.pressure(cv)
+        assert discr.norm(p - p_exact, np.inf) < tolerance
+        flux = inviscid_flux(discr, eos, cv)
+        logger.info(f"{dim}d flux = {flux}")
+        vel_exact = mom / mass
+
+        # first two components should be nonzero in livedim only
+        assert discr.norm(flux.mass - mom, np.inf) == 0
+        eflux_exact = (energy + p_exact)*vel_exact
+        assert discr.norm(flux.energy - eflux_exact, np.inf) == 0
```
Comment (inducer) [inline test/test_euler.py:255]: :tada: (OK to resolve when seen)

File: test/test_euler.py [hunk 8]
```diff
@@ -707,11 +667,11 @@ def test_multilump_rhs(actx_factory, dim, order, v0):
...
 
         inviscid_rhs = euler_operator(
-            discr, eos=IdealSingleGas(), boundaries=boundaries, q=lump_soln, t=0.0)
-        expected_rhs = lump.exact_rhs(discr, lump_soln, 0)
+            discr, eos=IdealSingleGas(), boundaries=boundaries, cv=lump_soln, t=0.0)
+        expected_rhs = lump.exact_rhs(discr, cv=lump_soln, t=0)
         print(f"inviscid_rhs = {inviscid_rhs}")
         print(f"expected_rhs = {expected_rhs}")
-        err_max = discr.norm(inviscid_rhs-expected_rhs, np.inf)
+        err_max = discr.norm((inviscid_rhs-expected_rhs).join(), np.inf)
```
Comment (inducer) [inline test/test_euler.py:674]: https://github.com/inducer/grudge/issues/97#issuecomment-858169109 (OK to resolve when seen)

File: test/test_init.py [hunk 9]
```diff
@@ -376,13 +367,11 @@ def test_multilump(ctx_factory, dim):
                               spec_centers=centers, velocity=velocity,
                               spec_y0s=spec_y0s, spec_amplitudes=spec_amplitudes)
 
-    lump_soln = lump(nodes)
-    numcvspec = get_num_species(dim, lump_soln)
+    cv = lump(nodes)
+    numcvspec = get_num_species(dim, cv.join())
```
Comment (inducer) [inline test/test_init.py:371]: Should we deprecate `get_num_species`? (OK to turn that into an issue.)

**Review by inducer — APPROVED** [review]
Thanks, this looks great! A few notes below, otherwise good to go. FWIW, this was pretty painless to review.

File: test/test_euler.py (hunk 7 above)
Comment (MTCam) [reply test/test_euler.py:255]: Indeed.

File: mirgecom/simutil.py (hunk 4 above)
Comment (MTCam) [reply mirgecom/simutil.py:115]: This isn't readily apparent unless you are really looking hard, but...

the broken-out `err_norms` is reported to the user so he can track the errors in each component of the solution.  

I do see the merit in replacing the simple `max` with a norm in isolation... since we need the components anyway, however, I'm not really sure.

fyi, when I actually do the replacement `maxerr=discr.norm(exp_resid, np.inf)` didn't work.  I had to: `maxerr=discr.norm(exp_resid.join(), np.inf)`.

File: test/test_init.py (hunk 9 above)
Comment (MTCam) [reply test/test_init.py:371]: What about making it internal-only?   Or would you like to make an `nspecies` property?

**Comment by MTCam** [comment]
Narrowly avoided a gitastrophy.
